pkg/platform/runtime/setup/setup.go

```diff
@@ -220,44 +220,27 @@ func (s *Setup) installArtifacts(buildResult *model.BuildResult, artifacts artif
 	if !buildResult.BuildReady && buildResult.BuildEngine == model.Camel {
 		return locale.NewInputError("build_status_in_progress", "", apimodel.ProjectURL(s.target.Owner(), s.target.Name(), s.target.CommitUUID().String()))
 	}
-	// Artifacts are installed in two stages
-	// - The first stage runs concurrently in MaxConcurrency worker threads (download, unpacking, relocation)
-	// - The second stage moves all files into its final destination is running in a single thread (using the mainthread library) to avoid file conflicts
-
-	// installErr is fed with exactly one error object (or nil pointer) after the installation go-routine finishes
-	installErr := make(chan error)
-
-	// artfErrs collects all errors that can happen in worker threads in
-	artfErrs := make(chan error)
-	go func() {
-		var errs []error
-		for err := range artfErrs {
-			errs = append(errs, err)
-		}
-
-		if len(errs) > 0 {
-			installErr <- &ArtifactSetupErrors{errs}
-		} else {
-			installErr <- nil
-		}
-	}()
 
+	installErrs := make(chan error)
 	// schedule the first stage, binding mainthread library to this thread
 	mainthread.Run(func() {
-		defer close(artfErrs)
-		var err error
+		defer close(installErrs)
 		if buildResult.BuildReady {
-			err = s.installFromBuildResult(buildResult, artifacts, artfErrs)
+			installErrs <- s.installFromBuildResult(buildResult, artifacts, installErrs)
 		} else {
-			err = s.installFromBuildLog(buildResult, artifacts, artfErrs)
-		}
-
-		if err != nil {
-			artfErrs <- err
+			installErrs <- s.installFromBuildLog(buildResult, artifacts, installErrs)
 		}
 	})
 
-	return <-installErr
+	var errs []error
+	for err := range installErrs {
+		errs = append(errs, err)
+	}
```

Comment:
I think we'll never get here, because the `installErrs` channel will be blocked inside the `Run()` method which then doesn't return.

Reply:
You can try it.  It stops forever on the same error, or when all artifacts are installed.

Reply:
Why is it blocked inside the `Run()` method?

Anyway, don't mean to turn your PR into my PR, key point here is I think we can improve the error handling here. The current PR has too much going on in it for my comfort..